src/pkg/artifact/dao/dao.go

```diff
@@ -251,131 +241,89 @@ func (d *dao) DeleteReferences(ctx context.Context, parentID int64) error {
 	return err
 }
 
-func querySetter(ctx context.Context, query *q.Query) (beegoorm.QuerySeter, error) {
-	qs, err := orm.QuerySetter(ctx, &Artifact{}, query)
-	if err != nil {
-		return nil, err
-	}
-	qs, err = setBaseQuery(qs, query)
-	if err != nil {
-		return nil, err
-	}
-	qs, err = setTagQuery(ctx, qs, query)
-	if err != nil {
-		return nil, err
-	}
-	qs, err = setLabelQuery(qs, query)
-	if err != nil {
-		return nil, err
-	}
-	return qs, nil
-}
-
 // handle q=base=*
 // when "q=base=*" is specified in the query, the base collection is the all artifacts of database,
 // otherwise the base connection is only the tagged artifacts and untagged artifacts that aren't
 // referenced by others
-func setBaseQuery(qs beegoorm.QuerySeter, query *q.Query) (beegoorm.QuerySeter, error) {
-	if query == nil || len(query.Keywords) == 0 {
-		qs = qs.FilterRaw("id", both)
-		return qs, nil
-	}
-	base, exist := query.Keywords["base"]
-	if !exist {
-		qs = qs.FilterRaw("id", both)
-		return qs, nil
-	}
-	b, ok := base.(string)
-	if !ok || b != "*" {
-		return qs, errors.New(nil).WithCode(errors.BadRequestCode).
-			WithMessage(`the value of "base" query can only be exact match value with "*"`)
+func filterBase(query *q.Query) selector.Option {
+	return func(builder *selector.Builder, cond *selector.Cond) error {
+		base, exist := query.GetValue("base")
+		if !exist {
+			builder.Where(
+				cond.Or(
+					cond.Exists("SELECT 1 FROM tag WHERE tag.artifact_id = T0.id"),
```

Comment:
Is possible to hide the `T0` which is the implementation detail of the `selector`?

Reply:
No, it's not possible to hide the `T0`. When applying `JOIN tag` and filter by `repository_id` to the select, the SQL doesn't know where the repository_id comes from.